cmd/csi-api-gen/generators/groups_and_versions.go

```diff
@@ -0,0 +1,227 @@
+	if !present {
+		klog.Fatalf("did not find interface %s in package %s", d.serverInterfaceName(), versionPkg.Path)
+	}
+	if serverInterface.Kind != types.Interface {
+		klog.Fatalf("type %s in package %s should be an interface, it actually is a %s",
+			d.serverInterfaceName(), versionPkg.Path, serverInterface.Kind)
+	}
+
+	version := &apiVersion{
+		Package: versionPkg,
+	}
+	d.versions = append(d.versions, version)
+
+	for callbackName, versionedCallback := range serverInterface.Methods {
+		d.validateServerCallback(callbackName, versionedCallback, version)
+
+		version.serverCallbacks.getOrInsert(namedCallback{
+			name:     callbackName,
+			callback: versionedCallback,
+		})
+
+		namedServerCallback := namedCallback{
+			name:     callbackName,
+			callback: replaceTypesPackage(versionedCallback, versionPkg.Path, pkgPlaceholder),
+		}
+
+		if previousCallback := d.serverCallbacks.getOrInsert(namedServerCallback); previousCallback != nil {
+			if namedServerCallback.callback.String() != previousCallback.callback.String() {
+				errorMsg := fmt.Sprintf("Endpoint %s in API group %s inconsistent across versions:", callbackName, d.name)
+				for _, vsn := range d.versions {
+					if vsnCallback := vsn.serverCallbacks.get(callbackName); vsnCallback != nil {
+						errorMsg += fmt.Sprintf("\n  - in version %s: %s", vsn.Name, vsnCallback.callback)
+					}
+				}
+				errorMsg += fmt.Sprintf("\nYields 2 different signatures for the internal server callback:\n%s\nand\n%s",
+					previousCallback, namedServerCallback.callback)
+				klog.Fatalf(errorMsg)
+			}
+		}
+	}
+}
+
+// validateServerCallback checks that server callbacks have the expected shape, i.e.:
+// * all versioned (i.e. in the same package) parameter should be pointers
+// * return values should all be pointers, except for the last one, that must be an error
+// These assumptions are
+func (d *groupDefinition) validateServerCallback(callbackName string, callback *types.Type, version *apiVersion) {
+	for _, param := range callback.Signature.Parameters {
+		if isVersionedVariable(param, version) && param.Kind != types.Pointer {
+			klog.Fatalf("Server callback %s in API %s version %s has a non-pointer versioned parameter: %v",
+				callbackName, d.name, version.Name, param)
+		}
+	}
+	for i, returnValue := range callback.Signature.Results {
+		if i == len(callback.Signature.Results)-1 {
+			if !isError(returnValue) {
+				klog.Fatalf("The last returned value for server callback %s in API %s version %s should be an error, found %v instead",
+					callbackName, d.name, version.Name, returnValue)
+			}
+		} else if returnValue.Kind != types.Pointer {
+			klog.Fatalf("Server callback %s in API %s version %s has a non-pointer return value: %v",
+				callbackName, d.name, version.Name, returnValue)
+		}
+	}
+}
+
+// isError returns true iff type t is the built-in type "error"
+func isError(t *types.Type) bool {
+	return t.Kind == types.Interface && t.Name.Name == "error" && t.Name.Package == ""
```

Comment:
The t.Name.Package must be empty string? What does it mean if it is not empty?

Reply:
It means that the type belongs to a package; that's not the case for built-in types.

Reply:
does it make sense to name it IsBuildinError to be more specific?

Reply:
Absolutely, changed, thanks :)